### DISCUSSION TITLE
Bug: require('wdio-electron-service/main') crashes app

### DISCUSSION THREAD

**Discussion by qwalley** [webdriverio-community/wdio-electron-service#413] [General]
Seems that vitest/spy does not enjoy being "require"d. Here's a minimal reproduction using the electron boilerplate project.

Environment: Ubuntu 22

Steps:
1. `git clone https://github.com/webdriverio/electron-boilerplate.git`
2. `cd electron-boilerplate && npm install && npm run package`
3. `NODE_ENV=test ./out/WebdriverIO\ +\ Electron\ Boilerplate-linux-x64/WebdriverIO\ +\ Electron\ Boilerplate`

Output:
```
A JavaScript error occurred in the main process
Uncaught Exception:
Error [ERR_REQUIRE_ESM]: require() of ES Module /home/walley/repos/electron-boilerplate/out/WebdriverIO + Electron Boilerplate-linux-x64/resources/app.asar/node_modules/@vitest/spy/dist/index.js from /home/walley/repos/electron-boilerplate/out/WebdriverIO + Electron Boilerplate-linux-x64/resources/app.asar/node_modules/wdio-electron-service/dist/cjs/main.js not supported.
Instead change the require of index.js in /home/walley/repos/electron-boilerplate/out/WebdriverIO + Electron Boilerplate-linux-x64/resources/app.asar/node_modules/wdio-electron-service/dist/cjs/main.js to a dynamic import() which is available in all CommonJS modules.
    at l._load (node:electron/js2c/asar_bundle:2:13642)
    at Object.<anonymous> (/home/walley/repos/electron-boilerplate/out/WebdriverIO + Electron Boilerplate-linux-x64/resources/app.asar/node_modules/wdio-electron-service/dist/cjs/main.js:30:15)
    at l._load (node:electron/js2c/asar_bundle:2:13642)
    at Object.<anonymous> (/home/walley/repos/electron-boilerplate/out/WebdriverIO + Electron Boilerplate-linux-x64/resources/app.asar/dist/app.js:17:5)
    at l._load (node:electron/js2c/asar_bundle:2:13642)
    at node:electron/js2c/browser_init:2:120247
    at node:electron/js2c/browser_init:2:120456
    at node:electron/js2c/browser_init:2:120460
    at l._load (node:electron/js2c/asar_bundle:2:13642)
```

**Comment by christian-bromann** [comment]
@qwalley I recommend to mock the `wdio-electron-service` dependency, there is no reason to actually import it into your application.

**Reply by qwalley** [reply]
@christian-bromann thanks for the response, I'm not sure that I understand why there is no need to import the service. I thought that the whole point of the `wdio-electron-service/main` and `wdio-electron-service/preload` modules was to expose the electron APIs and an ipc method for executing arbitrary javascript to the tests. Obviously in production these aren't wanted, but I thought that was the point of checking whether `NODE_ENV=test`.

**Reply by christian-bromann** [reply]
Based on your comment:

> Seems that vitest/spy does not enjoy being "require"d.

I assume you would import your application code in your unit tests. This is why I recommend to mock the package out since this service is not relevant for unit testing.

**Reply by christian-bromann** [reply]
> Steps:

I am not sure what this should reproduce here. You have to run `npm run package` to bundle the application. Also if you like to include the service code you have to set `NODE_ENV=test` during build time.

**Reply by qwalley** [reply]
Ah I see. No I'm trying to test my electron application with webdriverio and was having trouble. I was able to reproduce the problem by running the boilerplate project with `NODE_TEST=env`. The error message I get when running the boilerplate application indicates that there is a problem with the way `wdio-electron-service/main` is importing `@vitest/spy`.

**Reply by qwalley** [reply]
Ya sorry that was a typo in my post, the steps should be `npm run package` not `npm run build`

**Reply by christian-bromann** [reply]
Can you please provide a step by step direction to reproduce this?

**Reply by qwalley** [reply]
I provided steps in the post. Clone the boilerplate repo, run the npm package script, and then try running the application with `NODE_ENV=test`.

**Reply by christian-bromann** [reply]
I raised https://github.com/webdriverio/electron-boilerplate/issues/92

There are two things to this: the boilerplate doesn't bundle the application code and removes the `wdio-electron-service` if statement, so I raise above issue to resolve this. You should not bundle this code as part of your production environemnt.

Second, I am not totally sure why this there is an ESM/CJS issue with the bundled application and not in the test. Is there any reason why you start the application with `NODE_ENV=test`?

**Reply by qwalley** [reply]
Thanks for looking into this! 

I was only starting with `NODE_ENV=test` because I was trying to replicate the way chromedriver would be starting the application. When running `wdio` the `POST /session` request to the chromedriver server would timeout, but the chromedriver server wouldn't log any information about why. So I wanted to start the application manually to see if there were any errors, and what I found was the error above.

**Reply by christian-bromann** [reply]
I raised an issue. It seems like the CJS output target imports `@vitest/spy` which is ESM only. Not sure why this is.

**Reply by christian-bromann** [reply]
Ok, looking into the code we have introduced mocking capabilities for the Electron service and that seems to have broke the CJS support.

**Reply by goosewobbler** [reply]
I've been looking into this and these are my initial findings / thoughts:

* Step 2 of the reproduction doesn't pass for me - I get TS errors unable to find the main and preload scripts.
* I am not sure how the CJS E2Es of the Electron service are passing without this error occuring.  I even resurrected [wdio-electron-service-example](https://github.com/goosewobbler/wdio-electron-service-example) and updated it for the latest service version in order to test the E2Es in a separate repo using the service as installed from NPM - and they pass.  Perhaps this is something to do with the bundling and the issue only occurs when using unbundled source files - the service repo E2Es use rollup and `wdio-electron-service-example` uses webpack.
* `electron-boilerplate` is not explicitly tested as part of the service build process - perhaps we could use it to run the E2Es for the service (Docker?) - we would however need ESM, CJS, electron-builder variants.
* Supporting CJS is (still) a massive headache.  Feels like having to add loads of extra code and hacks for IE6 / IE8 back in the day...

I added a prospective fix PR [here](https://github.com/webdriverio-community/wdio-electron-service/pull/415), it passes the E2Es but so did the previous version...

**Reply by christian-bromann** [reply]
It makes sense to me now. The `/main` import does this `const spy_1 = require("@vitest/spy")` which doesn't work since `@vitest/spy` is an ESM package. I added a comment to the PR which asks if it is possible to dynamically import it or if there is a need to set `fn` as global.

**Reply by goosewobbler** [reply]
@qwalley Do you have reproduction steps for this which work?  `electron-boilerplate` is producing TS errors on the package step.

To test my hypothesis above, I'm currently looking to add yet another example app in the service, one which doesn't use a bundler - not managed to get it working yet.  Depending on how much work this ends up being, it may be worthwhile just telling people to use a bundler until we can get #308 built out.

**Reply by goosewobbler** [reply]
Never mind, I got `example-no-bundling` working and can now reproduce the error there.

**Reply by goosewobbler** [reply]
This should be fixed now in `6.3.1`.

**Reply by christian-bromann** [reply]
Thanks for your work @goosewobbler 🎉
